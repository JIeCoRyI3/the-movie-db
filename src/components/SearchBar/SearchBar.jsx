import React from 'react';
import styles from './SearchBar.module.css';
import {
    loadDataByGenre,
    loadDataByTitle,
    loadDataByGenreOrTitle,
} from './connect-store';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

export class SearchBar extends React.Component {
    state = {
        value: '',
    };

    searchByTitle = () => {
        this.pushGetParameters('query');
        this.props.loadDataByTitle(this.getUserInput('query'));
    };

    searchByGenre = () => {
        this.pushGetParameters('with_genres');
        this.props.loadDataByGenre(this.getUserInput('with_genres'));
    };

    searchByGenreOrTitle = () => {
        this.pushGetParameters('title_and_genres');
        this.props.loadDataByGenreOrTitle(
            this.getUserInput('title_and_genres')
        );
    };

    handleChange = (e) => {
        const value = e.target.value;
        this.setState({
            value,
        });
    };

    getUserInput = (searchBy) => {
        return {
            [searchBy]: this.state.value,
        };
    };

    pushGetParameters = (searchBy) => {
        const userInput = this.getUserInput(searchBy);
        const pushObj = {
            pathname: '/',
        };

        switch (searchBy) {
            case 'query':
                pushObj.search = `?searchBy=title${
                    userInput.query ? '&input=' + userInput.query : ''
                }`;
                break;
            case 'with_genres':
                pushObj.search = `?searchBy=genre${
                    userInput.with_genres
                        ? '&input=' + userInput.with_genres
                        : ''
                }`;
                break;
            case 'title_and_genres':
                pushObj.search = `?searchBy=genreOrTitle${
                    userInput.title_and_genres
                        ? '&input=' + userInput.title_and_genres
                        : ''
                }`;
                break;
            default:
                pushObj.search = `?searchBy=title${
                    userInput.query ? '&input=' + userInput.query : ''
                }`;
        }

        this.props.history.push(pushObj);
    };

    render() {
        return (
            <div className={styles.searchBarContainer}>
                <h4 className={styles.title}>Find your movies</h4>
                <input
                    className={styles.input}
                    placeholder="Enter movie title or genre"
                    id="searchInput"
                    onChange={this.handleChange}
                />
                <div className={styles.filterBarContainer}>
                    <div className={styles.leftSide}>
                        <h5
                            className={`${styles.leftSideItem} ${styles.searchTitle}`}
                        >
                            Search by
                        </h5>
                        <button
                            className={`${styles.leftSideItem} btn btn-primary`}
                            onClick={this.searchByTitle}
                            id="title"
                        >
                            Title
                        </button>
                        <button
                            className={`${styles.leftSideItem} btn btn-primary`}
                            onClick={this.searchByGenre}
                            id="genre"
                        >
                            Genre
                        </button>
                    </div>
                    <div>
                        <button
                            className={`btn btn-primary ${styles.rightSideItem}`}
                            onClick={this.searchByGenreOrTitle}
                            id="search"
                        >
                            SEARCH
                        </button>
                    </div>
                </div>
            </div>
        );
    }
}

SearchBar.displayName = 'SearchBar';

const mapDispatchToProps = {
    loadDataByGenre,
    loadDataByTitle,
    loadDataByGenreOrTitle,
};

const withStore = connect(null, mapDispatchToProps);

export default withRouter(withStore(SearchBar));
