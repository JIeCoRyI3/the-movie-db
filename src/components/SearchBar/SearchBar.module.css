.searchBarContainer {
    width: 90%;
    display: flex;
    flex-direction: column;
}

.title,
.searchTitle {
    font-family: Arial, Helvetica, sans-serif;
    font-style: italic;
    color: black;
    text-align: left;
    font-size: 2em;
    font-weight: bold;
    margin-right: 1em;
}

.filterBarContainer {
    width: 100%;
    display: flex;
    justify-content: space-between;
    margin-top: 1em;
    align-items: center;
}

.leftSide {
    display: flex;
    align-items: center;
}

.leftSideItem {
    margin-right: 1em;
    font-size: 1.4em;
    padding: 0.2em 1em;
}
.rightSideItem{
    font-size: 1.4em;
    padding: 0.2em 1em;
}

.input {
    margin-top: 1em;
    height: 2.8em;
    font-size: 1.4em;
    padding-left: 2em;
}
