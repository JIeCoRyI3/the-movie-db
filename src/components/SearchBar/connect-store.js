import { loadMovies } from '../../store/actions-creators/loadMovies';
import ApiClient from '../../api/apiClient';

const api = new ApiClient();

export const loadData = (filterObj) => {
    return (dispatch) => {
        const filter = localStorage.getItem('filter');

        if (filter === 'query') {
            api.filterBy(filterObj).then((res) => {
                dispatch(loadMovies(res.results));
            });
        } else if (filter === 'with_genres') {
            api.getAllGenres().then((genres) => {
                genres.genres.map((genre) => {
                    if (genre.name.includes(filterObj['with_genres'])) {
                        filterObj['with_genres'] = genre.id;
                    }
                    return 1;
                });

                api.filterWithGenres(filterObj).then((res) => {
                    dispatch(loadMovies(res.results));
                });
            });
        }
    };
};
