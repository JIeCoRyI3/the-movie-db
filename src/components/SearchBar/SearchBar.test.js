import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchBar from './SearchBar';

describe('<SearchBar/>', () => {
    it('should renders', () => {
        const component = shallow(<SearchBar />);
        expect(component).toMatchSnapshot();
    });

    // describe('searchByTitle', () => {
    //     it('should be called after click', () => {
    //         const component = mount(<SearchBar />);
    //         const instance = component.instance();
    //         jest.spyOn(instance, 'searchByTitle');
    //         expect(instance.searchByTitle).toHaveBeenCalled();
    //     });
    //
    // });
});
