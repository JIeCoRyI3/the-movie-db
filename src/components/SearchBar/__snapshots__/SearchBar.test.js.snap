// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar/> should render button 1`] = `
<button
  className="undefined btn btn-primary"
  id="title"
  onClick={[Function]}
>
  Title
</button>
`;

exports[`<SearchBar/> should render input 1`] = `
<input
  id="searchInput"
  onChange={[Function]}
  placeholder="Enter movie title or genre"
/>
`;

exports[`<SearchBar/> should renders 1`] = `
<div>
  <h4>
    Find your movies
  </h4>
  <input
    id="searchInput"
    onChange={[Function]}
    placeholder="Enter movie title or genre"
  />
  <div>
    <div>
      <h5
        className="undefined undefined"
      >
        Search by
      </h5>
      <button
        className="undefined btn btn-primary"
        id="title"
        onClick={[Function]}
      >
        Title
      </button>
      <button
        className="undefined btn btn-primary"
        id="genre"
        onClick={[Function]}
      >
        Genre
      </button>
    </div>
    <div>
      <button
        className="btn btn-primary undefined"
        id="search"
        onClick={[Function]}
      >
        SEARCH
      </button>
    </div>
  </div>
</div>
`;
